internal/certmonitor/certmonitor.go

```diff
@@ -0,0 +1,217 @@
+package certmonitor
+
+import (
+	"crypto/x509"
+	"encoding/base64"
+	"encoding/pem"
+	"fmt"
+	"github.com/stackrox/acs-fleet-manager/fleetshard/pkg/fleetshardmetrics"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/tools/cache"
+	"time"
+)
+
+// SelectorConfig represents a configuration to select a namespace or a secret by name or labelSelector. Only one of Name or LabelSelector can be specified.
+type SelectorConfig struct {
+	Name          string                `json:"name"`
+	LabelSelector *metav1.LabelSelector `json:"labelSelector"`
+}
+
+// MonitorConfig represents a configuration for observing certificates contained in kubernetes secrets
+type MonitorConfig struct {
+	Namespace SelectorConfig `json:"namespace"`
+	Secret    SelectorConfig `json:"secret"`
+}
+
+// Config represents the certificate monitor configuration
+
+type Config struct {
+	Monitors     []MonitorConfig `json:"monitors"`
+	ResyncPeriod *time.Duration  `json:"resyncPeriod"`
+}
+
+// NamespaceGetter interface for retrieving namespaces on name. This interface is a subset of `v1.NamespaceLister`
+type NamespaceGetter interface {
+	Get(name string) (*corev1.Namespace, error)
+}
+
+// certMonitor is the Certificate Monitor. It watches Kubernetes secrets containing certificates, and populates prometheus metrics with the expiration time of those certificates.
+type certMonitor struct {
+	informerfactory informers.SharedInformerFactory
+	podInformer     cache.SharedIndexInformer
+	config          *Config
+	namespaceGetter NamespaceGetter
+	metrics         *fleetshardmetrics.Metrics
+}
+
+// Start the certificate monitor
+func (c *certMonitor) Start(stopCh <-chan struct{}) error {
+	c.podInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    c.handleSecretCreation,
+		UpdateFunc: c.handleSecretUpdate,
+		DeleteFunc: c.handleSecretDeletion,
+	},
+	)
+	c.informerfactory.Start(stopCh)
+
+	if !cache.WaitForCacheSync(stopCh, c.podInformer.HasSynced) {
+		return fmt.Errorf("timed out waiting for caches to sync")
+	}
+	return nil
+}
+
+// NewCertMonitor creates new instance of certMonitor
+func NewCertMonitor(config *Config, informerFactory informers.SharedInformerFactory, podInformer cache.SharedIndexInformer, namespaceGetter NamespaceGetter) (*certMonitor, error) {
+	monitor := &certMonitor{
+		informerfactory: informerFactory,
+		podInformer:     podInformer,
+		config:          config,
+		namespaceGetter: namespaceGetter,
+		metrics:         fleetshardmetrics.MetricsInstance(),
+	}
+	return monitor, nil
+}
```

Comment:
This method never returns an error

```suggestion
func NewCertMonitor(config *Config, informerFactory informers.SharedInformerFactory, podInformer cache.SharedIndexInformer, namespaceGetter NamespaceGetter) *certMonitor {
	return &certMonitor{
		informerfactory: informerFactory,
		podInformer:     podInformer,
		config:          config,
		namespaceGetter: namespaceGetter,
		metrics:         fleetshardmetrics.MetricsInstance(),
	}
}
```